### ISSUE TITLE
Issue/Bug - Duolingo-Solution-Viewer doesn't except "i" as replacement for "i̇".

### ISSUE THREAD

**Issue by R0M4N0F5K7** [blmage/duolingo-solution-viewer#146]
This is a curious thing...
So sometimes when there are sentences with "i" in the correct answer in the Turkish Duolingo-course, the add-on counts it as incorrect character and displays the same sentence with "i"s with weird shifted dots as correct solution (picture below).
Apparently the "i" used by me is different than the "i̇" used for the answers written by the Duolingo Turkish staff. I suppose that the letter might be composed differently here software wise, because in Turkish "i"s with and without dots are different vowels.
This might be the code behind it: https://stackoverflow.com/questions/25995451/why-i-is-not-equals-to-i%CC%87
The weird floating dot that the solution viewer displays seems to be "Unicode Character 'COMBINING DOT ABOVE' (U+0307)".

So it seems like technically there are not one but two small issues:
1) the solution fewer doesn't convert the "i̇" made with i+floating dot to a single letter, or at least does not display it as such and
2) it does not, as it should, see "i" as equal alternative to "i̇" (which would be the desired behaviour, because then the solution viewer would not have to display the whole "corrected answer"-section in such instances)

I know that's not a huge deal, but I would appreciate a fix nonetheless.
Either way, thanks for this great add-on!

![Screenshot_20210222-222728~2](https://user-images.githubusercontent.com/15329506/109233987-23f11280-77cb-11eb-806d-f2c7922f99a5.png)

**Comment by blmage** [comment]
Thanks for the very detailed report!

There are three takeaways here:
- we should probably not correct answers when the word bank is used,
- there seems to be a problem with Unicode normalization in some cases,
- we should investigate whether the cleanup of solutions (especially w.r.t. diacritics) is relevant for Turkish solutions.

I'll try to gather a few different examples to help with the last two points, and update you on this shortly.

**Comment by R0M4N0F5K7** [comment]
Maybe I can help you with some more information on what kind of examples I found/had in the past:

>     * we should probably not correct answers when the word bank is used,
Maybe you had something in mind that I don't get because I am not familiar with the inner workings of the solution viewer, but just from what I found and the problem posted here: this issue appears independent from weather or not I use word bank or type by keyboard in duolingo, hence my suspicion that the weird "i" stems from the way some course contributors typed the "possible solution". Only sometimes the solution contains this weird "double dot-i", sometimes the "normal i", and if it is the "normal i", then the solution viewer recognizes it and doesn't output unwanted correction as it would do with "double dot-i"s.

>     * there seems to be a problem with Unicode normalization in some cases,
I am actually already halfway through the Turkish tree, and I only ever noticed it to have a problem with this "double dot-i" character exclusively if that helps. (The only reason I didn't report it before was, that I just started using the solution viewer add-on on firefox for android too, and there it takes up much more screen real estate when displaying those false corrections, which is a little bit annoying. On desktop it did not bother me so much.)

**Comment by R0M4N0F5K7** [comment]
> Hence, it seems that this would be another example of Duolingo generating incorrect variants of solutions (and would explain why there is a lot of redundancy in Turkish solutions containing `i`s) - could you confirm this?

To be honest, i have no idea. The more I read about it, the less I get where the different, sometimes weirdly displayed I with strange dot comes from. My original thought was, that it is just a different composition made by Turkish keyboards, as Turkish does distinguish between dotless i (I and ı) and dotted i (İ i), but every source I find says that it still has the same code, just the capital dotted i (İ) and the dotless i (ı) obviously have a different code.
My GUESS was, that it is an error on the "solution viewer"s end, because most browsers for example seem to display "i" (this is a regular i) and "i̇" (this is the weird i, that the solution viewer displays weirdly) identical, but then, weirdly enough, if I copy them and google them google puts out slightly different results for both characters. Firefox for Mobile is the only browser i tested that displays the two dotted i's differently (screenshot below).

But yeah, if you say that Duolingo has a habit of generating incorrect solutions, than I suppose it's on them. But either way, the simple solution is to make the solution viewer turn the weird i (i̇) into an regular i (i) and ignoring this differentiation. Maybe?

> Knowing that :
> 
>     * if it still is a valid character, I'll adapt the diff algorithm to try and better highlight where the difference is,
> 
>     * if it is an invalid character (making for invalid words), I'll make sure to clean up the Turkish solutions accordingly in the next version.

It seems to be a different charactar composition for what is supposed to be the same/identical and valid character = i. Just counting it as invalid character might be a solution, but I don't know if someone would maybe input their solutions with that weird i and then be corrected by the solution viewer? But I think you know better than I do - if the weird i-variants are just created by duolingo then the solution viewer should just see them as invalid character and not import them I guess. But I wonder why the solution viewer corrected my answers because of the different "i" in the first place, if there are also always variants with the proper "i" so that we could just scratch the solutions with the weird redundant "i"...

I'm sorry, I am afraid I am not of much help here. I think you actually got a better idea of what the problem is than I do by now.

__

I agree with your word bank observations and keeping the current behavior there.

**Comment by R0M4N0F5K7** [comment]
Sorry for replying on a closed topic, but I thought it might be good to keep the history of our observations of the "weird i" in one place.
First of all, I am amazed by your quick work and clean solution, seems to work great on SMALL i's, only in the rear cases when a CAPITAL i occurs things get messed up again (and weirdly affect other words too, but I suppose this would be fixed if the capital i's get the same exception-code(?) ). In the code (44e45ff) you commented that the capital i with dot is composed that way, but I don't think so, this character has it's individual code too (see [Dotted and dotless I on Wikipedia](https://en.wikipedia.org/wiki/Dotted_and_dotless_I))

TLDR: The code in 44e45ff (Filter out copies containing a "combining dot above" after a lowercase "i") should also be applied to uppercase "İ".

**Comment by blmage** [comment]
No problem, thanks for the update!

I think this could be related to some other character/invalid combination, given that the "İ" is correctly combined and should not cause any problem (also, the correction for "ispanyol" [seems appropriate](https://translate.google.com/?sl=en&tl=tr&text=Spanish&op=translate)) :
```javascript
>>> 'i\u0307'.normalize('NFC').length
2

>>> 'I\u0307'.normalize('NFC').length
1
```


The rest is much weirder though - could you point me to the lesson in which you got this example?

**Comment by R0M4N0F5K7** [comment]
If there isn't something I missed I'd still say that what I said is the issue. In your example below you only prove that I+dot is fine, but not İ+dot, don't you? 
> ```js
> 
> >>> 'I\u0307'.normalize('NFC').length
> 1
> ```
That's no proper analogy to the problem we had with the i+dot, as we had the problem with "i+dot" not with "ı+dot". So the issue we have with duolingo seems to occur when a i which already has a dot gets (for whatever reason) combined with a second dot. That's also why I linked the wikipedia-entry in my last post, to point out that the İ [capital i with dot] I'm talking about is an individual character (u0130). 
> the correction for "ispanyol" [seems appropriate](https://translate.google.com/?sl=en&tl=tr&text=Spanish&op=translate)
The correction (or missing correction] is appropriate. You gotta look closer at my screenshot - there you can see that there is a second dot attached next to the İ where I marked it, just in the same way that the i had two dots before you filtered it out and which was the initial issue of this post.
> The rest is much weirder though - could you point me to the lesson in which you got this example?
This was in the still "Nations", especially in the lesson 3/5 there are several country names with capital İ. Only in sentences with an İ we have those messed up characters appearing AFTER it. I have more screenshots that I didn't upload last time:
![Screenshot (187)](https://user-images.githubusercontent.com/15329506/111073155-efc16580-84dd-11eb-8a05-d0d962e7267b.png)
![Screenshot (186)](https://user-images.githubusercontent.com/15329506/111073162-f64fdd00-84dd-11eb-8ecb-a6a8e2b47cd2.png)
![Screenshot (183)](https://user-images.githubusercontent.com/15329506/111073193-167f9c00-84de-11eb-9335-4c08306d77e1.png)
![Screenshot (188)](https://user-images.githubusercontent.com/15329506/111073103-bd176d00-84dd-11eb-8da6-1b4208538a93.png)

**Comment by blmage** [comment]
Sorry, I can see better what's wrong now, thanks to your other screenshots.

The problem seems to come from the diff algorithm rather than from the solutions, because the difference between a `ı` and a `i` is a _combining dot above_.

This isolated _combining dot above_ could be what breaks the rest of the sentences too, if it messes with the calculation of offsets, or something like this.

I'll try to come up with a solution for the difference between `ı` and `i` first, and then we can advise depending on the effect it has on the rest.

**Comment by R0M4N0F5K7** [comment]
why don't we try adding

    result = result.filter(!/İ\u0307/.test(_));

or, if İ has to be replaced by javascript source code:

    result = result.filter(!/\u0130\u0307/.test(_));

? (I am horrible at js)

to 44e45ff? 

My thought is still, that the answers with the small i are perfect now after you added that code, only the captial İ has still the extra dot and messes stuff up. Whatever couses the weird character replacements afterwards, they must be still related to the captial İ displayed with one extra dot, so we might try adressing fixing the capital İ the same way we fixed the small i first.

**Comment by blmage** [comment]
Ok, I think I got it (at last)! 😄 

The problem came from the diff algorithm, which works on lowercase strings (to simplify the process).

A bad locale was used, so the `İ` was transformed into a `i` followed by a _combining dot above_, rather than just a `i`.

Note that the difference between "İspanyol" and "ispanyol" will be ignored from now on, as expected, because only the case varies.

Sorry for all the back and forth! this one proved to be more involved than I initially thought.
